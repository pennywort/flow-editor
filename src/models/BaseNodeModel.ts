import { Node } from '@xyflow/react';
import React from "react";

export interface NodeButton {
    label: string;
    target?: string;
    external?: boolean;
    href?: string;
}

export interface BaseNodeData extends Record<string, unknown> {
    label: string;
}

export class BaseNodeModel<T extends BaseNodeData = BaseNodeData> implements Node<T> {
    id: string;
    type: string;
    position: { x: number; y: number };
    data: T;
    style: React.CSSProperties;

    constructor(config: {
        id: string;
        type: string;
        position: { x: number; y: number };
        data: T;
        style?: React.CSSProperties;
    }) {
        this.id = config.id;
        this.type = config.type;
        this.position = config.position;
        this.data = config.data;
        this.style = this.getBaseStyle(config.style);
    }

    protected getBaseStyle(customStyles?: React.CSSProperties): React.CSSProperties {
        return {
            backgroundColor: 'rgb(0, 0, 0)',
            paddingLeft: '3px',
            color: '#fff',
            borderRadius: 8,
            width: 280,
            ...customStyles,
        };
    }
}