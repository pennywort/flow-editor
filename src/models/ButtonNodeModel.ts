import { BaseNodeModel, BaseNodeData, NodeButton } from './BaseNodeModel';

export type ButtonNodeData = {
    label: string;
    richText: string;
    buttons?: NodeButton[];
    expanded?: boolean | undefined;
} & BaseNodeData;

export class ButtonNodeModel extends BaseNodeModel<ButtonNodeData> {
	constructor(
		id: string,
		label: string,
		position: { x: number; y: number },
		content: string,
		buttons: NodeButton[] = [],
		extra?: Partial<Omit<ButtonNodeData, 'label' | 'richText' | 'buttons'>>
	) {
		super({
			id,
			type: 'textWithButtons',
			position,
			data: { label, richText: content, buttons, ...(extra || {}) },
		});
	}
}