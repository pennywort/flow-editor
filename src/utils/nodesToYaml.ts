import yaml from 'yaml';
import { Node } from '@xyflow/react';

import { NodeButton } from '../models/BaseNodeModel';
import { ButtonNodeData } from '../models/ButtonNodeModel';

interface YamlMessage {
    id: string;
    label: string;
    text?: string;
    child_messages?: string[];
    parent_message?: string;
    is_link_button?: boolean;
    back_label?: string;
    back_to_start_message_label?: string;
    delete_id?: string;
    delete_label?: string;
}

interface YamlRoot {
    id: string;
    text: string;
    back_label?: string;
    back_to_start_message_label?: string;
    delete_id?: string;
    delete_label?: string;
    child_messages: string[];
    messages: YamlMessage[];
}

export function nodesToYaml(nodes: Node<ButtonNodeData>[], rootId = 'menu'): string {
	const rootNode = nodes.find(n => n.id === rootId) || nodes[0];
	if (!rootNode) {
		throw new Error('Root node not found!');
	}

	// Сериализация кнопок
	function splitButtons(buttons: NodeButton[], parentId: string) {
		const child_message_ids: string[] = [];
		const link_buttons: YamlMessage[] = [];
		for (const btn of buttons) {
			if (btn.external) {
				link_buttons.push({
					id: btn.id,
					label: btn.label,
					parent_message: parentId,
					is_link_button: true
				});
			} else if (btn.target) {
				child_message_ids.push(btn.target);
			}
		}
		return { child_message_ids, link_buttons };
	}

	// --- Сборка YAML ---
	const messages: YamlMessage[] = [];
	let rootChildMessages: string[] = [];
	let linkButtons: YamlMessage[] = [];

	// Сначала root
	if (rootNode.data.buttons) {
		const split = splitButtons(rootNode.data.buttons, rootNode.id);
		rootChildMessages = split.child_message_ids;
		linkButtons = [...split.link_buttons];
	}

	nodes.forEach(node => {
		if (node.id === rootNode.id) {
			return;
		}
		const data = node.data;
		let child_messages: string[] = [];

		if (data.buttons) {
			const split = splitButtons(data.buttons, node.id);
			child_messages = split.child_message_ids;
			linkButtons.push(...split.link_buttons);
		}
		messages.push({
			id: node.id,
			label: data.label,
			text: data.richText,
			child_messages: child_messages.length ? child_messages : undefined,
			parent_message: (data as any).parent_message,
			back_label: (data as any).back_label,
			back_to_start_message_label: (data as any).back_to_start_message_label,
			delete_id: (data as any).delete_id,
			delete_label: (data as any).delete_label,
		});
	});

	// Добавляем все external кнопки в messages!
	messages.push(...linkButtons);

	const yamlObj: YamlRoot = {
		id: rootNode.id,
		text: rootNode.data.richText,
		back_label: (rootNode.data as any).back_label,
		back_to_start_message_label: (rootNode.data as any).back_to_start_message_label,
		delete_id: (rootNode.data as any).delete_id,
		delete_label: (rootNode.data as any).delete_label,
		child_messages: rootChildMessages,
		messages,
	};

	return yaml.stringify(yamlObj);
}
