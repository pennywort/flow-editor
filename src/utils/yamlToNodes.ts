import yaml from 'yaml';
import { Node } from '@xyflow/react';
import { ButtonNodeModel, ButtonNodeData } from '../models/ButtonNodeModel';
import { NodeButton } from '../models/BaseNodeModel';

export function yamlToNodes(yamlString: string): Node<ButtonNodeData>[] {
    const obj = yaml.parse(yamlString);

    const nodes: Node<ButtonNodeData>[] = [];
    const messageMap = new Map<string, any>();
    const linkButtons: any[] = [];

    // Разделим обычные сообщения и external link-кнопки
    (obj.messages ?? []).forEach((msg: any) => {
        if (msg.is_link_button) linkButtons.push(msg);
        messageMap.set(msg.id, msg);
    });

    function collectButtons(message: any): NodeButton[] {
        const result: NodeButton[] = [];

        // 1. Обычные дочерние кнопки (по child_messages)
        if (message.child_messages) {
            for (const childId of message.child_messages) {
                const child = messageMap.get(childId);
                if (child && !child.is_link_button) {
                    result.push({ label: child.label, target: childId, id: child.id });
                }
            }
        }

        // 2. External link-кнопки по parent_message
        for (const link of linkButtons) {
            if (link.parent_message === message.id) {
                result.push({
                    label: link.label,
                    external: true,
                    id: link.id,
                });
            }
        }
        return result;
    }

    // Root node
    nodes.push(
        new ButtonNodeModel(
            obj.id,
            '', // label у рутового блока может быть пустым
            { x: 50, y: 50 },
            obj.text,
            collectButtons(obj)
        )
    );

    (obj.messages ?? []).forEach((msg: any, i: number) => {
        if (msg.is_link_button) return; // сами external-кнопки как отдельные блоки не нужны!
        nodes.push(
            new ButtonNodeModel(
                msg.id,
                msg.label,
                { x: 300 + i * 40, y: 50 + i * 100 },
                msg.text,
                collectButtons(msg)
            )
        );
    });

    return nodes;
}
