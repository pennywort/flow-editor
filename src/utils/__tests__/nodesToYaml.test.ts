import { ButtonNodeModel } from '../../models/ButtonNodeModel';
import { yamlToNodes } from '../yamlToNodes';
import { nodesToYaml } from '../nodesToYaml';

const nodes = [
	new ButtonNodeModel(
		'menu',
		'Меню',
		{ x: 50, y: 50 },
		'Привет, я бот - навигатор по базе знаний "Сервисы ИБ Фабрики Данных".',
		[
			{ label: 'Категория 1', target: 'first_category', id: '123' },
			{ label: 'Категория 2', target: 'second_category', id: '321' },
		]
	),
	new ButtonNodeModel(
		'first_category',
		'Категория 1',
		{ x: 300, y: 50 },
		'Текст категории 1',
		[
			{ label: 'Перейти в Категорию 2', target: 'first_button_1', id: '5453' },
		]
	),
	new ButtonNodeModel(
		'second_category',
		'Категория 2',
		{ x: 300, y: 200 },
		'Текст категории 2',
		[
			{ label: 'Внешний ресурс', external: true, id: 'link_test' },
		]
	),
	new ButtonNodeModel(
		'first_button_1',
		'Перейти в Категорию 2',
		{ x: 600, y: 50 },
		'...',
		[]
	),
];

describe('utils test', () => {
	it('конвертирует nodes в yaml (snapshot)', () => {
		const yamlStr = nodesToYaml(nodes, 'menu');
		expect(yamlStr).toMatchSnapshot('nodes-to-yaml');
	});

	it('конвертирует yaml в nodes (snapshot)', () => {
		const yamlStr = nodesToYaml(nodes, 'menu');
		const nodes2 = yamlToNodes(yamlStr);
		expect(nodes2).toMatchSnapshot('yaml-to-nodes');
	});
});


describe('roundtrip conversion', () => {
	it('nodes → yaml → nodes2 → yaml2 сохраняет структуру', () => {
		const yaml1 = nodesToYaml(nodes, 'menu');
		const nodes2 = yamlToNodes(yaml1);
		const yaml2 = nodesToYaml(nodes2, 'menu');
		expect(nodes2).toMatchSnapshot('nodes2-after-roundtrip');
		expect(yaml2).toMatchSnapshot('yaml2-after-roundtrip');
	});
});
