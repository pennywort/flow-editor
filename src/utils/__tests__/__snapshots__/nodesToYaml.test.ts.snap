// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`roundtrip conversion nodes → yaml → nodes2 → yaml2 сохраняет структуру: nodes2-after-roundtrip 1`] = `
Array [
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "external": false,
          "label": "Категория 1",
          "target": "first_category",
        },
        Object {
          "external": false,
          "label": "Категория 2",
          "target": "second_category",
        },
      ],
      "label": "",
      "richText": "Привет, я бот - навигатор по базе знаний \\"Сервисы ИБ Фабрики Данных\\".",
    },
    "id": "menu",
    "position": Object {
      "x": 50,
      "y": 50,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "external": false,
          "label": "Перейти в Категорию 2",
          "target": "first_button_1",
        },
      ],
      "label": "Категория 1",
      "richText": "Текст категории 1",
    },
    "id": "first_category",
    "position": Object {
      "x": 300,
      "y": 50,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "external": true,
          "href": "https://example.com",
          "label": "Внешний ресурс",
        },
      ],
      "label": "Категория 2",
      "richText": "Текст категории 2",
    },
    "id": "second_category",
    "position": Object {
      "x": 340,
      "y": 150,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [],
      "label": "Перейти в Категорию 2",
      "richText": "...",
    },
    "id": "first_button_1",
    "position": Object {
      "x": 420,
      "y": 350,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
]
`;

exports[`roundtrip conversion nodes → yaml → nodes2 → yaml2 сохраняет структуру: yaml2-after-roundtrip 1`] = `
"id: menu
text: Привет, я бот - навигатор по базе знаний \\"Сервисы ИБ Фабрики Данных\\".
back_label: Назад
delete_id: delete_menu
delete_label: Отмена
child_messages:
  - first_category
  - second_category
messages:
  - id: first_category
    label: Категория 1
    text: Текст категории 1
    child_messages:
      - first_button_1
  - id: second_category
    label: Категория 2
    text: Текст категории 2
  - id: \\"\\"
    label: \\"[Внешний ресурс](https://example.com)\\"
    text: \\"\\"
    parent_message: second_category
    is_link_button: true
  - id: first_button_1
    label: Перейти в Категорию 2
    text: ...
"
`;

exports[`utils test конвертирует nodes в yaml (snapshot): nodes-to-yaml 1`] = `
"id: menu
text: Привет, я бот - навигатор по базе знаний \\"Сервисы ИБ Фабрики Данных\\".
child_messages:
  - first_category
  - second_category
messages:
  - id: first_category
    label: Категория 1
    text: Текст категории 1
    child_messages:
      - first_button_1
  - id: second_category
    label: Категория 2
    text: Текст категории 2
  - id: first_button_1
    label: Перейти в Категорию 2
    text: ...
  - id: link_test
    label: Внешний ресурс
    parent_message: second_category
    is_link_button: true
"
`;

exports[`utils test конвертирует yaml в nodes (snapshot): yaml-to-nodes 1`] = `
Array [
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "id": "first_category",
          "label": "Категория 1",
          "target": "first_category",
        },
        Object {
          "id": "second_category",
          "label": "Категория 2",
          "target": "second_category",
        },
      ],
      "label": "",
      "richText": "Привет, я бот - навигатор по базе знаний \\"Сервисы ИБ Фабрики Данных\\".",
    },
    "id": "menu",
    "position": Object {
      "x": 50,
      "y": 50,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "id": "first_button_1",
          "label": "Перейти в Категорию 2",
          "target": "first_button_1",
        },
      ],
      "label": "Категория 1",
      "richText": "Текст категории 1",
    },
    "id": "first_category",
    "position": Object {
      "x": 300,
      "y": 50,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [
        Object {
          "external": true,
          "id": "link_test",
          "label": "Внешний ресурс",
        },
      ],
      "label": "Категория 2",
      "richText": "Текст категории 2",
    },
    "id": "second_category",
    "position": Object {
      "x": 340,
      "y": 150,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
  ButtonNodeModel {
    "data": Object {
      "buttons": Array [],
      "label": "Перейти в Категорию 2",
      "richText": "...",
    },
    "id": "first_button_1",
    "position": Object {
      "x": 380,
      "y": 250,
    },
    "style": Object {
      "backgroundColor": "rgb(0, 0, 0)",
      "borderRadius": 8,
      "color": "#fff",
      "paddingLeft": "3px",
      "width": 280,
    },
    "type": "textWithButtons",
  },
]
`;
